Created a PySide2 and Shiboken2 cmake finder.

ezzieyguywuf opened the pull request:
Thank you for creating a pull request to contribute to FreeCAD! To ease integration, please confirm the following:

- [ x] Branch rebased on latest master `git pull --rebase upstream master`
- [ ] Unit tests confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [ x] Commit message is [well-written](https://chris.beams.io/posts/git-commit/)
- [ n/a] Commit message includes `issue #<id>` or `fixes #<id>` where `<id>` is the [associated MantisBT](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) issue id if one exists

And please remember to update the Wiki with the features added or changed once this PR is merged.

---

wwmayer commented:
Maybe I don't understand the logic but why are PYSIDE2_FOUND and SHIBOKEN2_FOUND set to TRUE if the prior call of execute_process was not successful?
And what's the point of removing SHIBOKEN_INCLUDE_DIR and PYSIDE_INCLUDE_DIR?
Is it the plan to use shiboken's/PySide's Python instead of its C++ interface to wrap and unwrap objects?

A regression of this PR is that all travis builds are broken now.

ezzieyguywuf commented:
Thank you for reviewing this pull request. As it stands, I can see that travis builds are failing: I will look into why this is and update my pull request when I have resolved it. In the mean time, you can feel free to place this on hold, or I can delete the request altogether.

To address your questions:

    1. The result is 0 on success, 1 on failure (as is the case with all behaving processes). In CMake, the [if command](https://cmake.org/cmake/help/latest/command/if.html) evaluates a 1 as true and a 0 as false, thus the seemingly bad logic here.
    2. Regarding the include directories, I thought that the pip package did not install these but I was wrong. Further, I thought that the FreeCAD project was not making use of the include directories, but this could be the reason the builds are failing. I'll probably start my investigation here.

wwmayer commented:
As far as I can see building the binaries works but it then raises an unexpected exception when using the PySide stuff internally: Gui.getMainWindow()

slokhorst commented:
I get the following warnings:
```
CMake Warning:
  Value of PYSIDE2_LIBRARY contained a newline; truncating

CMake Warning:
  Value of SHIBOKEN2_LIBRARY contained a newline; truncating

-- Generating done
CMake Warning:
  Value of PYSIDE2_LIBRARY contained a newline; truncating

CMake Warning:
  Value of SHIBOKEN2_LIBRARY contained a newline; truncating
```

ezzieyguywuf commented:
I've merged the latest master and updated the variable names to make more sense. I've also added some variables that it looks like FreeCAD cmake config calls for.

I'm uploading these changes to see if it helps at all with Travis-ci. I'm working on rebuilding Qt on my comp, so I'm struggling to check the changes.

ezzieyguywuf commented:
I have succesfully run most of the freecad tests using the latest commit.

I cannot run all of the tests due to a local (personal) issue involving not being able to obtain a working copy of pivy.

However, as far as I can tell, all the pyside2 and shiboken stuff work fine.

I am not sure why there are file conflicts in my merge, as my commit has integrated the latest master. I am open to ideas/suggestions.

ezzieyguywuf commented:
I fixed the file conflicts. It was due to git idiocy on my part.

looooo commented:
As this is a major change to the way how we are trying to find shiboken2 / pyside2 please let me know once this PR is ready to check if it is also working with conda. In my eyes the config files coming with shiboken2 / pyside2 were not too bad (at least for virtual envs).

looooo commented on cMake/FindShiboken2.cmake:
```diff
@@ -0,0 +1,29 @@
+set(SHIBOKEN2_FOUND FALSE)
+if(NOT FAILURE)
+    set(SHIBOKEN2_FOUND TRUE)
+    set(SHIBOKEN2_LIBRARY ${PRINT_OUTPUT} CACHE PATH "The location of the Shiboken2 library")
```
I am really not sure if this is really what we want here. I guess we need "libshiboken" instead of the python library. Have a look at the filestructure of the conda pyside2-package:

```
pyside2-5.9.0a1-py37h4dc837a_0/
├── lib
│   ├── cmake
│   │   ├── PySide2-5.9.0
│   │   │   ├── PySide2Config.cmake
│   │   │   ├── PySide2Config.cpython-37m-x86_64-linux-gnu.cmake
│   │   │   └── PySide2ConfigVersion.cmake
│   │   └── Shiboken2-5.9.0
│   │       ├── Shiboken2Config.cmake
│   │       ├── Shiboken2Config.cpython-37m-x86_64-linux-gnu.cmake
│   │       └── Shiboken2ConfigVersion.cmake
│   ├── libpyside2.cpython-37m-x86_64-linux-gnu.so.5.9.0
│   ├── libpysidetest.so
│   ├── libshiboken2.cpython-37m-x86_64-linux-gnu.so.5.9.0
│   └── python3.7
│       └── site-packages
│           ├── PySide2
│           │   ├── _config.py
│           │   ├── _git_pyside_version.py
│           │   ├── __init__.py
│           │   ├── __pycache__
│           │   │   ├── _config.cpython-37.pyc
│           │   │   ├── _git_pyside_version.cpython-37.pyc
│           │   │   └── __init__.cpython-37.pyc
│           │   └── ...
│           ├── pyside2uic
│           │   └── ...
│           └── shiboken2.cpython-37m-x86_64-linux-gnu.so
└── ...
```
The path we derive by importing the shiboken2 via python will be something like `.../lib/pythonx.y/site-packages/shiboken2*.so` but actually we need `.../lib/libshiboken2*.so`. Maybe there is a method to derive this path via the shiboken python module, but I am not sure. I guess there will be a similar problem for pyside2 / libpyside2...

wwmayer commented:
#2477 is merged now which invalidates this PR. So it must be reworked.

ezzieyguywuf commented:
Good point, I will do that

El El jue, sep. 26, 2019 a las 5:45 p. m., wwmayer <notifications@github.com>
escribió:

> #2477 <https://github.com/FreeCAD/FreeCAD/pull/2477> is merged now which
> invalidates this PR. So it must be reworked.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/FreeCAD/FreeCAD/pull/2020?email_source=notifications&email_token=AAEMB5G2PBANNZXCEL6FNXTQLUURFA5CNFSM4G67ZFY2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7XBVRI#issuecomment-535698117>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEMB5BIJZNETQFF4D3Q3TLQLUURFANCNFSM4G67ZFYQ>
> .
>

slokhorst commented:
PySide2 and Shiboken2 are now properly detected on my machine, maybe because of 3939e4543ef4be6c50581d9bff77e930fb1d9d03 ?

looooo commented:
I guess this is how we should link to these libraries:
https://github.com/FreeCAD/FreeCAD/pull/2639/commits/b624d759ff7122a5c3b34526b90de44a7bd5d532

This solved pyside2/shiboken2 linking for Linux/Mac/windows (conda).

slokhorst commented:
I'm not sure whats going on.. I need to manually add `-DPYSIDE_LIBRARY="/usr/lib/libpyside2.cpython-37m-x86_64-linux-gnu.so"` to the `cmake` command, otherwise the linking fails with the previously mentioned error.

slokhorst commented:
I just compiled FreeCAD again without any PySide-specific options, and it did fine! This is the first time I tried this after 7222b32cdeae078305d319ef11b6574e701372a0 so maybe that did the trick?

@ezzieyguywuf can you compile master now without this PR? Or are there still issues finding PySide2 and Shiboken2?

softmoth commented:
@slokhorst, I was able to compile without the PySide2 options. FreeCAD master (git rev f47c6539b844b189b632303edfb4358e70acf74a).

ezzieyguywuf commented:
I haven’t done dev work on freecad in a few months, and I’m currently
focusing on other projects.

Therefore I will not be able to check this in the near future

El El jue, abr. 23, 2020 a la(s) 1:02 p. m., Tim Siegel <
notifications@github.com> escribió:

> @slokhorst <https://github.com/slokhorst>, I was able to compile without
> the PySide2 options. FreeCAD master (git rev f47c653
> <https://github.com/FreeCAD/FreeCAD/commit/f47c6539b844b189b632303edfb4358e70acf74a>
> ).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/FreeCAD/FreeCAD/pull/2020#issuecomment-618519052>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEMB5BORHSGJET2I34MQVTROBYA5ANCNFSM4G67ZFYQ>
> .
>

slokhorst commented:
@softmoth thanks for the feedback.

Maybe we should close this PR and revisit in case other users encounter compilation issues?
